scripts/usb_scan.py

```diff
@@ -0,0 +1,748 @@
+#!/usr/bin/env python
+#
+# Copyright (C) Citrix Systems Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published
+# by the Free Software Foundation; version 2.1 only. #
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# package dependency: python-pyudev
+#
+# called by XAPI to do
+# 1. a "lsusb" operation: get all usb device and interface information
+# 2. check if device can be passed through based on policy file
+# 3. return the device info to XAPI in json format
+
+import abc
+import argparse
+import json
+import xcp.logger as log
+import logging
+import pyudev
+import re
+import traceback
+
+# True to print log to user.log
+__DIAGNOSTIC = False
+
+
+def hex_equal(h1, h2):
+    """ check if the value of hex string are equal
+
+    :param h1:(str) lhs hex string
+    :param h2:(str) rhs hex string
+    :return: bool, if equal
+    """
+    try:
+        return int(h1, 16) == int(h2, 16)
+    except (TypeError, ValueError):
+        return False
+
+
+class UsbObject(dict):
+    """ Base class of USB classes, save USB properties in dict
+
+    node(str): the key, device node
+    """
+    __metaclass__ = abc.ABCMeta
```

Comment:
From a design perspective, why did you decide to use abstract base classes instead of piping the data to filters and fill a dictionary?

Reply:
This UsbObject is a dict.
I just want to put a few common methods in it.